### DISCUSSION TITLE
Filtering a collection for the table layout using url parameters

### DISCUSSION THREAD

**Discussion by mark-james** [directus/directus#6493] [Archive]
Hi,
I'm looking to create an internal project management app for my business leveraging the Directus admin app. I'd like to be able to access all the tasks related to a project via a button on the project detail screen.

Eg on /admin/collections/projects/1 use a button that links to /admin/collections/tasks/?filter[project.id]=1

Is there a way to apply this kind of filter to the table layout dynamically rather than relying on bookmarks?

Thanks
Mark

**Comment by benhaynes** [comment]
Ideally you'd want to create a O2M field from project to task. This would show a listing of related tasks that can open in a drawer when clicked.

**Reply by mark-james** [reply]
Thanks Ben. 

I am using an O2M field. Assuming I'm configuring things correctly, the drawer display templates work well for a small number of child records. But functionality-wise they are a fair way off the table layout to display larger record counts.

So currently there is no way of going to a filtered list of tasks using the table layout that haven't been preconfigured by a bookmark?

Thanks
Mark

**Reply by benhaynes** [reply]
I don't think there's a way to link to a Collection with filter params in the URL, that said, this is a great feature and one we'd definitely like to have!

**Reply by mark-james** [reply]
Ok, thanks Ben. 

It would be good to understand what's involved to put this change in. Happy to put together something if it's a relatively straightforward PR. But I will need some direction.

I guess the alternative route is to offer an alternative table display type to embed in the collections detail view (as a alternative to the drawers). But my assumption is that would be a more difficult change?

**Reply by benhaynes** [reply]
I think both of those are great options worth adding in time. Here is how I imagine both working:

**Collection Detail Params** — ability to append any [general query parameters](https://docs.directus.io/reference/api/query/) to the Collection Detail URL. Though I'm not sure if we can do limit/offset... we might want to only use `page`. @rijkvanzanten can advise on how/where this would be implemented.

**Table Listing Interface** — This could be a O2M (table style), but I think the current O2M (list style) already covers the relationship pretty well. Instead, I think it would be more useful to add an alias interface that simply pulls in any items based on a filter in the Interface Options. In that way, you can pull in any relevant items without needing to create a formal relationship.

**Reply by mark-james** [reply]
I agree, both would be good.

The first option - Collection Detail Params - gets me closer to the experience I'm after with tasks management. So I'd like to focus on that, assuming it's within my abilities. @rijkvanzanten - Can you please let me know what's involved?

**Reply by rijkvanzanten** [reply]
Heya!

This is quite a tricky one to implement, as we'll have to sync the filters between the URL and the internal state of the page/layout live, without messing with the browser history too much. There's one more issue: not every query param is available for every layout. limit/offset/page only really applies to table/cards, but wouldn't have any effect on map/calendar, likewise, advanced filters are augmented in calendar/map to only show items applicable to the current view in the layout. We'll have to do some more research on what parts can go in the URL, and how that should be handled across layouts. 

@nickrum This also ties into our conversation the other day about what part of the query state is managed on the page vs layout level

**Reply by christianrr** [reply]
I would need a similar feature. My use case are events with guests, with events having a O2M relationship to guests (hundreds of them). As @mark-james noted, the O2M interface is greatly flexible, but unhandy when it comes to large datasets, sorting and filtering. My first idea was to include a presentation link that triggers a filtered table layout with a query param (filter guests by event). But I understand that this raises a lot of other challenges.

Wouldn't it be a possibility though to extend the O2M interface with a button "Display in table layout" or something similar, which then shows the table layout with the filtered dataset (only items of the O2M relationship). This could be an option in the O2M interface to allow for toggling that, since it only makes sense for larger datasets where filtering is involved.

In the table layout, however, it would raise the need for an option to reset the "subset filter" to show all items again. Maybe the filter interface could be used for that!?

**Reply by benhaynes** [reply]
I think it would be best to build pagination into the relational interfaces... and/or show the first X and then have a button to "Show All" in the drawer (which uses the Layout and already has pagination). Thoughts on this approach?

**Reply by christianrr** [reply]
Yeah, right. I think the approach with showing only a certain amount of items (definable via interface options) and having a "Show All" button would be fairly easy to implement and would fit absolutely my needs...

"Show all" could then open the existing drawer with a title like "Browse items" and no select checkboxes. The drawer in its current form features layout, pagination, sorting and filtering... So everything which is needed basically. Clicking an item in the drawer could open the detail view of the content item, as the relational interface does already.

Totally overlooked that, great solution! ❤️

One other side note: I had several use cases where I would have wanted that the relational interface opens the nested item fullscreen, not in the drawer. Would it make sense to make the "target" for opening items in relational interfaces configurable?

**Reply by benhaynes** [reply]
Nice! For the "Open Item" thing, I think it might be best to add an "Open as Page" button to all item drawers. That way, you can easily open in a drawer (without losing your location or staged changes) but have the option to open the full item page from any drawer. It's TWO steps instead of one, but requires less configuration and works for all drawers.

**Reply by christianrr** [reply]
Right, staged changes are tricky in that context. Opening an item as page would always require to discard any staged changes, right?

**Reply by benhaynes** [reply]
Discard or save, yeah. Eventually we will have staged changes save to localstorage... but not yet.

**Comment by mark-james** [comment]
Hi @rijkvanzanten , @benhaynes 
Thanks for the detailed response. I'm still keen to make the chang, but I appreciate it's something I won't be able to push through quickly. Lots of things to consider. Do you want me to raise either an issue to start tracking it? Or keep it here for now?

I'd like to get both your feedback on what I'm trying to do to see if there is a simpler short term solution for my use case. I'm building out an OKR mvp and timing is pretty important. It's going to be used by executives, they aren't power users.

My current issue is drilling through from Objective to Key Result to Tasks isn't a particularly smooth experience when going a few relationships deep. Forms that a great for data entry aren't as 'nice' for browsing the data. The list O2M style is very flexible but having a large number of tasks in a list is quite hard to work with. Each line is a different length, you can't sort, filter etc. A future opportunity for inline editing probably isn't possible? Are there features of the list O2M that I am not appreciating that would help with this? Because it does appear that a new table interface component addresses a different type of usecase / data to the list. 

Could you suggest another way of structuring / accessing / displaying this task data with current capabilities the suit browsing vs editing?

Another feature that may be nice is to differentiate between the form that creates a new item and one that updates an item. During creation, things like the name, description make sense to be first, however during updating, it's more likely to be fields like status, assignee etc that will need to be updated. I think having a unique layout for both would improve the experience. 

Thoughts?

**Reply by benhaynes** [reply]
Just noting that you can show/hide/disable fields in create versus edit, but rearranging them isn't possible as of now. Not sure that's 80/20 either.

**Reply by rijkvanzanten** [reply]
> Do you want me to raise either an issue to start tracking it? Or keep it here for now?

Lets keep discussing implementation details here until we reach a final plan which we can execute 🙂 

> isn't a particularly smooth experience when going a few relationships deep

It's a little hard for me to imagine/visualize the exact flow you're working with just based on the above description alone 🤔  Would it be possible to share a database dump of a minimal setup you're trying to achieve, so we're all looking at/discussing the same thing? Always down to discuss ways to improve Directus! Sounds like there's some low-hanging fruit that can greatly improve things on the relational interface side.

As for the difference between create and update, I get where you're coming from, but that creates a whole lot of overhead for a feature that might be a bit too "power user only" 🤔

**Comment by Pachat** [comment]
I face the same problem. Lets capitalize on the project / tasks of the OP. Each tasks has a key referring to a project's id.
A button in every project would open the list of the tasks related to this project. 

_It's possible to have a **hundred** tasks for a project. Each task has a status, a manager, due dates..._

In such a condition, the advantage of having **a filtered collection** by the project's id is that, on the filtered collection of tasks, 
the user may
- navigate on multi-pages
- add new filters (e.g. on status, manager, due dates...)
- and sort the tasks by any column

The URL could be like  
/admin/collections/tasks/?filter={"id":{"_eq":"**`id`**"}} 
where **`id`** is obtained by clicking in the drop down list of the fields after clicking on the **`+`** at the right of the `URL` field of the button link.

Sort of preset bookmark but with the value of the project's `id` not "hard coded" in the preset filter but passed to it in the front-end depending on the opened project.

**Comment by antran22** [comment]
I am designing a system where there is a One Blog Post - Many Comments relationship, and Admin can Approve or Disapprove comments. The number of Comments on a Post can be large, so listing them all in a One-to-Many field in a Post detail page can be very inefficient. I would like to create a Button Link on the Post detail page, which will direct Admin to the Comments list page, and set the filtering field appropriately.

I think the way to go for this is to store and infer the filtering on the List page directly from URL query.

**Reply by benhaynes** [reply]
You can try the "Button Link" interface. It might allow you to build a dynamic URL to link to the desired page.

**Comment by u12206050** [comment]
Since my [discussion](https://github.com/directus/directus/discussions/16228) on this topic and PR got closed I guess I'll settle to support this :) It isn't a must have, but would make have made onboarding users a lot easier I believe. Now we have resorted to our own custom modules since the content module is too complicated for non technical users.

**Reply by yakubovmax213** [reply]
@u12206050 Hi, do you have a code example of how you made a custom module that allows filtering items by url?

**Reply by u12206050** [reply]
I implemented it on the native content module: https://github.com/directus/directus/pull/14081

**Reply by yakubovmax213** [reply]
Am I understanding correctly that the Directus team did not accept your changes? And I can take the code from your PR to apply this to my custom module?
